### ISSUE TITLE
Setting a new key could (?optionally?) relegate old key to index

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#2888]
Consider

```
DT = data.table(
  ID1 = rep(10:1, 10:1),
  ID2 = rep(1:10, 1:10)
)
setkey(DT, ID1)
# \/ could lead to ID1 becoming an index \/
setkey(DT, ID2)
```

Besides the storage cost of storing the sort order, adding `ID1` as an index during `setkey(DT, ID2)` is basically free. 

Can be useful if you're hopping through various sort orders in your work flow.

**Comment by jangorecki** [comment]
Changing key is not that common, at least it shouldn't be. Key column(s) should depends on data, not operations you are performing on data. The latter is better to be handled by indices. I can image it isn't optimized yet everywhere so setting key might give benefits now. Anyway the more indices on data the better, especially when they come for free.

**Comment by MichaelChirico** [comment]
I agree. My use case is i switched to a new key for convenience during plotting/looping over groups based on a non-primary key. table is small enough (roughly 500K) that this is all for clean syntax rather than speed per se.

**Comment by MichaelChirico** [comment]
Worked on this a little bit:

```
  if (length(o)) {
    if (verbose) {
      tt = system.time(.Call(Creorder,x,o))
      cat("reorder took", tt["user.self"]+tt["sys.self"], "sec\n")
    } else {
      .Call(Creorder,x,o)
    }
    # o gives the ordering under the new key
    #   with respect to the old key -- can assign
    #   this as an index directly
    if (!is.null(old_key) && old_key_to_index) {
      setattr(x, "index", integer())
      setattr(attr(x, "index", exact=TRUE),
              paste0("__", old_key, collapse=""), o)
    }
  }
```

Actually it strikes me that it might be better to make the argument as `keep_indices` instead, and apply `set_attr(attr(x, 'index'), old_index_i, attr(attr(x, 'index'), old_index_i)[o])` to all existing indices, as well as assigning `o` as its own index. 

Will come back to this.

**Comment by jangorecki** [comment]
Existing indices has to be re-ordered according to new key (as in your examples `.[o]`). And possibly, not sure, it may be cheaper to do it for all at once, and not for every index separately.
I'm not sure if this is not a duplicate of some existing issue.

**Comment by MichaelChirico** [comment]
maybe adding everything columnwise into a matrix and reordering is the way
to go. IIRC there's something in the comments about "maybe retain indices
in the future" but no issue #

**Comment by MichaelChirico** [comment]
@jangorecki  indeed I think this issue has converged on #1158, which is yours.
